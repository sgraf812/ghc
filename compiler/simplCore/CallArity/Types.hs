module CallArity.Types where

import BasicTypes
import CoreSyn
import Id
import Maybes
import Outputable
import UnVarGraph
import Usage
import VarEnv

import Control.Monad ( guard )
import Data.List ( tails, partition )

---------------------------------------
-- Functions related to UsageType    --
---------------------------------------

-- Result type for the two analyses.
-- See Note [Analysis I: The arity analyis]
-- and Note [Analysis II: The Co-Called analysis]
data UsageType
  = UT
  { ut_cocalled :: !UnVarGraph
  -- ^ Models cardinality, e.g. at most {1, many} via the co-call relation
  , ut_uses :: !(VarEnv SingleUse)
  -- ^ Models how an Id was used, if at all
  , ut_args :: !UsageSig
  -- ^ Collects the signature for captured lambda binders
  , ut_stable :: Bool
  -- ^ Entirely irrelevant for Usage information, but needed for detecting
  -- stabilization of fixed-point iteration.
  }

modifyArgs :: (UsageSig -> UsageSig) -> UsageType -> UsageType
modifyArgs modifier ut = ut { ut_args = modifier (ut_args ut) }

modifyCoCalls :: (UnVarGraph -> UnVarGraph) -> UsageType -> UsageType
modifyCoCalls modifier ut = ut { ut_cocalled = modifier (ut_cocalled ut) }

-- | How an expression uses its interesting variables
-- and the elaborated expression with annotated Ids
type AnalResult = (UsageType, CoreExpr)

emptyUsageType :: UsageType
emptyUsageType = UT emptyUnVarGraph emptyVarEnv topUsageSig False

botUsageType :: UsageType
botUsageType = unusedArgs emptyUsageType

unitUsageType :: Id -> SingleUse -> UsageType
unitUsageType id use = emptyUsageType { ut_uses = unitVarEnv id use }

unusedArgs :: UsageType -> UsageType
unusedArgs ut = ut { ut_args = botUsageSig }

delUsageTypes :: [Id] -> UsageType -> UsageType
delUsageTypes ids ae = foldr delUsageType ae ids

delUsageType :: Id -> UsageType -> UsageType
delUsageType id (UT g ae args s) = UT (g `delNode` id) (ae `delVarEnv` id) args s

domType :: UsageType -> UnVarSet
domType ut = varEnvDom (ut_uses ut)

domTypes :: [UsageType] -> UnVarSet
domTypes = foldr unionUnVarSet emptyUnVarSet . map domType

makeIdArg :: Id -> UsageType -> UsageType
makeIdArg id ut = delUsageType id (modifyArgs (consUsageSig (lookupUsage NonRecursive ut id)) ut)

-- In the result, find out the minimum arity and whether the variable is called
-- at most once.
lookupUsage :: RecFlag -> UsageType -> Id -> Usage
lookupUsage rec (UT g ae _ _) id = case lookupVarEnv ae id of
  Just use
    -- we assume recursive bindings to be called multiple times, what's the
    -- point otherwise? It's a little sad we don't encode it in the co-call
    -- graph directly, though.
    -- See Note [Thunks in recursive groups]
    | isRec rec -> manifyUsage (Used Once use)
    | id `elemUnVarSet` neighbors g id -> Used Many use
    | otherwise -> Used Once use
  Nothing -> botUsage

calledWith :: UsageType -> Id -> UnVarSet
calledWith ut id = neighbors (ut_cocalled ut) id

-- Replaces the co-call graph by a complete graph (i.e. no information)
multiplyUsages :: Multiplicity -> UsageType -> UsageType
multiplyUsages Once ut = ut
multiplyUsages Many ut@(UT _ u args _)
  = UT
  { ut_cocalled = completeGraph (domType ut)
  , ut_uses = mapVarEnv (\use -> bothSingleUse use use) u
  , ut_args = manifyUsageSig args
  , ut_stable = False
  }

asCompleteGraph :: UsageType -> Maybe UnVarSet
asCompleteGraph ut = do
  s <- UnVarGraph.isCompleteGraph_maybe (ut_cocalled ut)
  guard (sizeUnVarSet s == sizeUnVarSet (domType ut))
  return s

bothUsageTypes :: [UsageType] -> UsageType
bothUsageTypes uts 
  = UT cocalled uses args False
  where
    uses = foldr bothUseEnv emptyVarEnv (map ut_uses uts)
    args 
      = ut_args 
      . expectJust "bothUsageTypes: argument list empty" 
      . listToMaybe
      $ uts
    -- Cross calls between complete graphs yield another complete graph.
    -- This is an important optimization for large tuples, like occuring
    -- when analysing top-level binds as let bindings
    (completes, incompletes) 
      -- = ([], uts)
      = partition (isJust . asCompleteGraph) uts
    completeDom
      = unionUnVarSets 
      . map (fromJust . asCompleteGraph) 
      $ completes
    completePart = completeGraph completeDom
    graphs = completePart : map ut_cocalled incompletes
    doms = completeDom : map domType incompletes
    -- all tails, beginning with a two element list
    pairings = drop 2 . reverse . tails $ doms
    crossCalls = flip map pairings $ \(dom:others) -> 
      -- There room for improvement by reusing the result of `unionUnVarSets`
      completeBipartiteGraph dom (unionUnVarSets others) 
    cocalled = unionUnVarGraphs (graphs ++ crossCalls)

-- | Corresponds to sequential composition of expressions.
-- Used for application and cases.
-- Note this returns the @UsageSig@ from the first argument.
bothUsageType :: UsageType -> UsageType -> UsageType
bothUsageType ut1 ut2 = bothUsageTypes [ut1, ut2]

-- | Used when combining results from alternative cases
lubUsageType :: UsageType -> UsageType -> UsageType
lubUsageType (UT g1 u1 args1 _) (UT g2 u2 args2 _)
  = UT
  { ut_cocalled = unionUnVarGraph g1 g2
  , ut_uses = lubUseEnv u1 u2
  , ut_args = lubUsageSig args1 args2
  , ut_stable = False
  }

lubUseEnv :: VarEnv SingleUse -> VarEnv SingleUse -> VarEnv SingleUse
lubUseEnv = plusVarEnv_C lubSingleUse

bothUseEnv :: VarEnv SingleUse -> VarEnv SingleUse -> VarEnv SingleUse
bothUseEnv = plusVarEnv_C bothSingleUse

lubUsageTypes :: [UsageType] -> UsageType
lubUsageTypes = foldl lubUsageType botUsageType

-- | Peels off a single argument usage from the signature, corresponding to how
-- @App f a@ uses @a@ under the given incoming arity.
peelArgUsage :: UsageType -> (Usage, UsageType)
peelArgUsage ut = (usg, ut { ut_args = args' })
  where
    (usg, args') = unconsUsageSig (ut_args ut)

markStable :: UsageType -> UsageType
markStable ut = ut { ut_stable = True }

-- * Pretty-printing

instance Outputable UsageType where
  ppr (UT cocalled arities args _) = vcat
    [ text "arg usages:" <+> ppr args
    , text "co-calls:" <+> ppr cocalled
    , text "uses:" <+> ppr arities
    ]
