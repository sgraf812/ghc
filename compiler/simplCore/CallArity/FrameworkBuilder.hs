{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module CallArity.FrameworkBuilder
  ( FrameworkNode
  , TransferFunction
  , ChangeDetector
  , Worklist.alwaysChangeDetector
  , DataFlowFramework
  , FrameworkBuilder
  , RequestedPriority (..)
  , registerTransferFunction
  , dependOnWithDefault
  , buildAndRun
  ) where

import CallArity.Types
import Outputable
import Usage
import qualified Worklist

import Data.IntMap.Lazy (IntMap)
import qualified Data.IntMap.Lazy as IntMap
import Data.Map.Strict   (Map)
import qualified Data.Map.Strict as Map
import qualified Data.Set as Set
import Data.Maybe
import Control.Monad.Fix
import Control.Monad.Trans.State.Strict

newtype FrameworkNode
  = FrameworkNode Int
  deriving (Show, Eq, Ord, Outputable)

type TransferFunction a = Worklist.TransferFunction (FrameworkNode, SingleUse) AnalResult a
type ChangeDetector = Worklist.ChangeDetector (FrameworkNode, SingleUse) AnalResult
type DataFlowFramework = Worklist.DataFlowFramework (FrameworkNode, SingleUse) AnalResult
-- | Maps @FrameworkNode@ to incoming usage dependent @TransferFunction@s
type NodeTransferEnv = IntMap (SingleUse -> TransferFunction AnalResult, ChangeDetector)

newtype FrameworkBuilder a
  = FB { unFB :: State NodeTransferEnv a }
  deriving (Functor, Applicative, Monad)

buildFramework :: FrameworkBuilder a -> (a, DataFlowFramework)
buildFramework (FB state) = (res, Worklist.DFF dff)
  where
    (res, env) = runState state IntMap.empty -- NodeTransferEnv
    dff (FrameworkNode node, use) = case IntMap.lookup node env of
      Nothing -> pprPanic "CallArity.FrameworkBuilder.buildFramework" (ppr node)
      Just (transfer, detectChange) -> (transfer use, detectChange)

data RequestedPriority
  = LowerThan !FrameworkNode
  | HighestAvailable

registerTransferFunction
  :: RequestedPriority
  -> (FrameworkNode -> FrameworkBuilder (a, (SingleUse -> TransferFunction AnalResult, ChangeDetector)))
  -> FrameworkBuilder a
registerTransferFunction prio f = FB $ do
  nodes <- get
  let node = case prio of
        HighestAvailable -> 2 * IntMap.size nodes
        LowerThan (FrameworkNode node)
          | not (IntMap.member (node - 1) nodes) -> node - 1
          | otherwise -> pprPanic
            "CallArity.FrameworkBuilder.registerTransferFunction"
            (text "There was already a node registered with priority" <+> ppr (node - 1))
  (result, _) <- mfix $ \ ~(_, entry) -> do
    -- Using mfix so that we can spare an unnecessary Int counter in the state.
    -- Also because @f@ needs to see its own node in order to define its
    -- transfer function in case of letrec.
    modify' (IntMap.insert node entry)
    unFB (f (FrameworkNode node))
  return result

dependOnWithDefault :: AnalResult -> (FrameworkNode, SingleUse) -> TransferFunction AnalResult
dependOnWithDefault def which = do
  --which <- pprTrace "dependOnWithDefault:before" (ppr which) (return which)
  res <- fromMaybe def <$> Worklist.dependOn which
  --res <- pprTrace "dependOnWithDefault:after " (ppr which) (return res)
  return res

buildAndRun :: FrameworkBuilder (SingleUse -> TransferFunction AnalResult) -> SingleUse -> AnalResult
buildAndRun buildTransfer use = lookup_result (Worklist.runFramework fw (Set.singleton (node, use)))
  where
    (node, fw) = buildFramework $
      registerTransferFunction (LowerThan (FrameworkNode 0)) $ \node -> do
        transfer <- buildTransfer
        return (node, (transfer, Worklist.alwaysChangeDetector))

    lookup_result :: Map (FrameworkNode, SingleUse) AnalResult -> AnalResult
    lookup_result result_map = case Map.lookup (node, use) result_map of
      Nothing -> pprPanic "CallArity.FrameworkBuilder.buildAndRun" empty
      Just res -> res
