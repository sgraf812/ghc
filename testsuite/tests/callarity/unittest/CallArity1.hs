{-# LANGUAGE TupleSections #-}
import BasicTypes
import CoreSyn
import CoreUtils
import Id
import Type
import MkCore
import CallArity.Analysis ( callArityRHS )
import MkId
import SysTools
import DynFlags
import ErrUtils
import Outputable
import TysWiredIn
import Literal
import GHC
import Control.Monad
import Control.Monad.IO.Class
import System.Environment ( getArgs )
import VarSet
import PprCore
import Unique
import UniqSet
import CoreLint
import FastString
import FamInstEnv
import Name

-- Build IDs. use mkTemplateLocal, more predictable than proper uniques
go, go2, x, d, n, y, z, p, _1, _2 :: Id
[go,go2, x, d, n, y, z, p, _1, _2] = mkLocalTestIds
    (words "go go2 x d n y z p _1 _2")
    [ mkFunTys [intTy, intTy] intTy
    , mkFunTys [intTy, intTy] intTy
    , intTy
    , mkFunTys [intTy] intTy
    , mkFunTys [intTy] intTy
    , intTy
    , intTy
    , pairType -- implicitly bound to a case binder when matching a pair
    , mkFunTys [intTy] intTy -- fst pair selector, implicitly bound in case match
    , mkFunTys [intTy] intTy -- snd pair selector
    ]

f :: Id -- protoypical external function, for which there is no signature info
f = mkGlobalTestId 42 "f" (mkFunTys [intTy, intTy] intTy)

pairType :: Type
pairType = mkBoxedTupleTy [mkFunTys [intTy] intTy, mkFunTys [intTy] intTy]

exprs :: [(String, CoreExpr)]
exprs =
  [ ("go2",) $ --pprTraceIt "go2" $
     mkRFun go [x]
        (mkNrLet d (mkACase (Var go `mkVarApps` [x])
                          (mkLams [y] $ Var y)
                  ) $ mkLams [z] $ Var d `mkVarApps` [x]) $
        go `mkLApps` [0, 0]
  , ("nested_go2",) $
     mkRFun go [x]
        (mkNrLet n (mkACase (Var go `mkVarApps` [x])
                          (mkLams [y] $ Var y))  $
            mkACase (Var n) $
                mkFun go2 [y]
                    (mkNrLet d
                        (mkACase (Var go `mkVarApps` [x])
                                 (mkLams [y] $ Var y) ) $
                        mkLams [z] $ Var d `mkVarApps` [x] )$
                    Var go2 `mkApps` [mkLit 1] ) $
        go `mkLApps` [0, 0]
  , ("d0 (go _*C^1(C^1(U)) would be bad)",) $
     mkRFun go [x]
        (mkNrLet d (mkACase (Var go `mkVarApps` [x])
                          (mkLams [y] $ Var y)
                  ) $
            mkLams [z] $ Var f `mkApps` [ Var d `mkVarApps` [x],  Var d `mkVarApps` [x] ]) $
        go `mkLApps` [0, 0]
  , ("go2 (in case crut)",) $
     mkRFun go [x]
        (mkNrLet d (mkACase (Var go `mkVarApps` [x])
                          (mkLams [y] $ Var y)
                  ) $ mkLams [z] $ Var d `mkVarApps` [x]) $
        Case (go `mkLApps` [0, 0]) z intTy
            [(DEFAULT, [], Var f `mkVarApps` [z,z])]
  , ("go2 (in function call)",) $
     mkRFun go [x]
        (mkNrLet d (mkACase (Var go `mkVarApps` [x])
                          (mkLams [y] $ Var y)
                  ) $ mkLams [z] $ Var d `mkVarApps` [x]) $
        f `mkLApps` [0] `mkApps` [go `mkLApps` [0, 0]]
  , ("go2 (using surrounding interesting let)",) $
     mkNrLet n (f `mkLApps` [0]) $
         mkRFun go [x]
            (mkNrLet d (mkACase (Var go `mkVarApps` [x])
                              (mkLams [y] $ Var y)
                      ) $ mkLams [z] $ Var d `mkVarApps` [x]) $
            Var f `mkApps` [n `mkLApps` [0],  go `mkLApps` [0, 0]]
  , ("two calls, one from let and from body (d 1*_ would be bad)",) $
     mkNrLet  d (mkACase (mkLams [y] $ Var y) (mkLams [y] $ Var y)) $
     mkFun go [x,y] (mkVarApps (Var d) [x]) $
     mkApps (Var d) [mkLApps go [1,2]]
  , ("a thunk in a recursion (d 1*_ would be bad)",) $
     mkRLet n (mkACase (mkLams [y] $ Var y) (Var n)) $
     mkRLet d (mkACase (mkLams [y] $ Var y) (Var d)) $
         Var n `mkApps` [d `mkLApps` [0]]
  , ("two thunks, one called multiple times (both 1*_ would be bad!)",) $
     mkNrLet n (mkACase (mkLams [y] $ mkLit 0) (f `mkLApps` [0])) $
     mkNrLet d (mkACase (mkLams [y] $ mkLit 0) (f `mkLApps` [0])) $
         Var n `mkApps` [Var d `mkApps` [Var d `mkApps` [mkLit 0]]]
  , ("two functions, not thunks",) $
     mkNrLet go  (mkLams [x] (mkACase (mkLams [y] $ mkLit 0) (Var f `mkVarApps` [x]))) $
     mkNrLet go2 (mkLams [x] (mkACase (mkLams [y] $ mkLit 0) (Var f `mkVarApps` [x]))) $
         Var go `mkApps` [go2 `mkLApps` [0,1], mkLit 0]
  , ("a thunk, called multiple times via a forking recursion (d 1*_ would be bad!)",) $
     mkNrLet  d   (mkACase (mkLams [y] $ mkLit 0) (f `mkLApps` [0])) $
     mkRLet go2 (mkLams [x] (mkACase (Var go2 `mkApps` [Var go2 `mkApps` [mkLit 0, mkLit 0]]) (Var d))) $
         go2 `mkLApps` [0,1]
  , ("a function, one called multiple times via a forking recursion",) $
     mkNrLet go   (mkLams [x] (mkACase (mkLams [y] $ mkLit 0) (Var f `mkVarApps` [x]))) $
     mkRLet go2 (mkLams [x] (mkACase (Var go2 `mkApps` [Var go2 `mkApps` [mkLit 0, mkLit 0]]) (go `mkLApps` [0]))) $
         go2 `mkLApps` [0,1]
  , ("two functions (recursive)",) $
     mkRLet go  (mkLams [x] (mkACase (Var f `mkVarApps` [x]) (Var go `mkVarApps` [x]))) $
     mkRLet go2 (mkLams [x] (mkACase (Var f `mkVarApps` [x]) (Var go2 `mkVarApps` [x]))) $
         Var go `mkApps` [go2 `mkLApps` [0,1], mkLit 0]
  , ("mutual recursion (thunks), called mutiple times (both 1*_ would be bad!)",) $
     Let (Rec [ (n, mkACase (mkLams [y] $ mkLit 0) (Var d))
              , (d, mkACase (mkLams [y] $ mkLit 0) (Var n))]) $
         Var n `mkApps` [Var d `mkApps` [Var d `mkApps` [mkLit 0]]]
  , ("mutual recursion (functions), but no thunks",) $
     Let (Rec [ (go,  mkLams [x] (mkACase (mkLams [y] $ mkLit 0) (Var go2 `mkVarApps` [x])))
              , (go2, mkLams [x] (mkACase (mkLams [y] $ mkLit 0) (Var go `mkVarApps` [x])))]) $
         Var go `mkApps` [go2 `mkLApps` [0,1], mkLit 0]
  , ("mutual recursion (functions), one boring (d 1*_ would be bad)",) $
     mkNrLet d (f `mkLApps` [0]) $
         Let (Rec [ (go,  mkLams [x, y] (Var d `mkApps` [go2 `mkLApps` [1,2]]))
                  , (go2, mkLams [x] (mkACase (mkLams [y] $ mkLit 0) (Var go `mkVarApps` [x])))]) $
             Var d `mkApps` [go2 `mkLApps` [0,1]]
  , ("a thunk (non-function-type), called twice, still calls once",) $
    mkNrLet d (f `mkLApps` [0]) $
        mkNrLet x (d `mkLApps` [1]) $
            Var f `mkVarApps` [x, x]
  , ("a thunk (function type), called multiple times, still calls once",) $
    mkNrLet d (f `mkLApps` [0]) $
        mkNrLet n (Var f `mkApps` [d `mkLApps` [1]]) $
            mkLams [x] $ Var n `mkVarApps` [x]
  , ("a thunk (non-function-type), in mutual recursion, still calls once (d w*_ would be bad)",) $
    mkNrLet d (f `mkLApps` [0]) $
        Let (Rec [ (x, Var d `mkApps` [go `mkLApps` [1,2]])
                 , (go, mkLams [y] $ mkACase (mkLams [z] $ Var x) (Var go `mkVarApps` [x]) ) ]) $
            Var go `mkApps` [mkLit 0, go `mkLApps` [0,1]]
  , ("a thunk (non-function-type), in mutual recursion, causes many calls (d 1*_ would be bad)",) $
    mkNrLet d (f `mkLApps` [0]) $
        Let (Rec [ (x, Var go `mkApps` [go `mkLApps` [1,2], go `mkLApps` [1,2]])
                 , (go, mkLams [y] $ mkACase (Var d) (Var go `mkVarApps` [x]) ) ]) $
            Var go `mkApps` [mkLit 0, go `mkLApps` [0,1]]
  , ("a thunk (function type), in mutual recursion, still calls once",) $
    mkNrLet d (f `mkLApps` [0]) $
        Let (Rec [ (n, Var go `mkApps` [d `mkLApps` [1]])
                 , (go, mkLams [x] $ mkACase (Var f `mkVarApps` [x]) (Var go `mkApps` [Var n `mkVarApps` [x]]) ) ]) $
            Var go `mkApps` [mkLit 0, go `mkLApps` [0,1]]
  , ("a thunk (function type), in mutual recursion, absent",) $
    mkLet d (f `mkLApps` [0]) $
        Let (Rec [ (n, Var go `mkApps` [d `mkLApps` [1]]) -- FIXME: Check UsageSigs
                 , (go, mkLams [x] $ mkACase (Var n) (Var go `mkApps` [Var n `mkVarApps` [x]]) ) ]) $
            Var go `mkApps` [mkLit 0, go `mkLApps` [0,1]]
  , ("a thunk (non-function-type) co-calls with the body (d 1*_ would be bad)",) $
    mkNrLet d (f `mkLApps` [0]) $
        mkLet x (d `mkLApps` [1]) $
            Var d `mkVarApps` [x]
  , ("body cocalls d and n, n calls d (anything other than d w*C^w(U) = w*U would be bad)",) $
    mkNrLet d (f `mkLApps` [0]) $
        mkLet n (mkLams [y] $ d `mkLApps` [1]) $
            Var f `mkApps` [d `mkLApps` [0], n `mkLApps` [0]]
  , ("body calls d and n mutually exclusive, n calls d. d should be called once",) $
    mkNrLet d (f `mkLApps` [0]) $
        mkLet n (mkLams [y] $ d `mkLApps` [1]) $
            mkACase (d `mkLApps` [0]) (n `mkLApps` [0])
  -- Product related tests
  , ("calling the first tuple component once",) $
    mkLet d (f `mkLApps` [0]) $
        mkLet n (mkLams [y] $ d `mkLApps` [1]) $
            elimPair (mkVarPair d n) (_1 `mkLApps` [0])
  , ("calling the second tuple component twice (expect n 1*U and d w*U by transitivity)",) $
    mkLet d (f `mkLApps` [0]) $
        mkLet n (mkLams [y] $ d `mkLApps` [1]) $
            elimPair (mkVarPair d n) (Var _2 `mkApps` [_2 `mkLApps` [0]])
  ]

main = do
    [libdir] <- getArgs
    runGhc (Just libdir) $ do
        getSessionDynFlags >>= setSessionDynFlags . flip gopt_set Opt_SuppressUniques
        dflags <- getSessionDynFlags
        liftIO $ forM_ exprs $ \(n,e) -> do
            case lintExpr dflags [f] e of
                Just msg -> putMsg dflags (msg $$ text "in" <+> text n)
                Nothing -> return ()
            putMsg dflags (text n <> char ':')
            -- liftIO $ putMsg dflags (ppr e)
            let e' = callArityRHS dflags emptyFamInstEnvs e
            let bndrs = nonDetEltsUniqSet (allBoundIds e')
              -- It should be OK to use nonDetEltsUniqSet here, if it becomes a
              -- problem we should use DVarSet
            -- liftIO $ putMsg dflags (ppr e')
            forM_ bndrs $ \v -> putMsg dflags $ nest 4 $ ppr v <+> ppr (idCallArity v)

-- Utilities
mkLApps :: Id -> [Integer] -> CoreExpr
mkLApps v = mkApps (Var v) . map mkLit

mkACase = mkIfThenElse (Var trueDataConId)

mkLocalTestId :: Int -> String -> Type -> Id
mkLocalTestId i s ty = mkSysLocal (mkFastString s) (mkBuiltinUnique i) ty

mkLocalTestIds :: [String] -> [Type] -> [Id]
mkLocalTestIds ns tys = zipWith3 mkLocalTestId [0..] ns tys

mkGlobalTestId :: Int -> String -> Type -> Id
mkGlobalTestId i s ty = mkVanillaGlobal (mkSystemVarName (mkBuiltinUnique i) (mkFastString s)) ty

mkNrLet :: Id -> CoreExpr -> CoreExpr -> CoreExpr
mkNrLet v rhs body = Let (NonRec v rhs) body

mkRLet :: Id -> CoreExpr -> CoreExpr -> CoreExpr
mkRLet v rhs body = Let (Rec [(v, rhs)]) body

mkFun :: Id -> [Id] -> CoreExpr -> CoreExpr -> CoreExpr
mkFun v xs rhs body = mkNrLet v (mkLams xs rhs) body

mkRFun :: Id -> [Id] -> CoreExpr -> CoreExpr -> CoreExpr
mkRFun v xs rhs body = mkRLet v (mkLams xs rhs) body

mkLit :: Integer -> CoreExpr
mkLit i = Lit (mkLitInteger i intTy)

pairDataCon :: DataCon
pairDataCon = tupleDataCon Boxed 2

mkPair :: CoreExpr -> CoreExpr -> CoreExpr
mkPair fst snd = mkCoreConApps pairDataCon
  [ Type (CoreUtils.exprType fst)
  , Type (CoreUtils.exprType snd)
  , fst
  , snd
  ]

elimPair :: CoreExpr -> CoreExpr -> CoreExpr
elimPair pair alt
  = Case pair p (CoreUtils.exprType alt) [(DataAlt pairDataCon, [_1, _2], alt)]

mkVarPair :: Id -> Id -> CoreExpr
mkVarPair fst snd = mkPair (Var fst) (Var snd)

-- Collects all let-bound IDs
allBoundIds :: CoreExpr -> VarSet
allBoundIds (Let (NonRec v rhs) body) = allBoundIds rhs `unionVarSet` allBoundIds body `extendVarSet` v
allBoundIds (Let (Rec binds) body) =
    allBoundIds body `unionVarSet` unionVarSets
        [ allBoundIds rhs `extendVarSet` v | (v, rhs) <- binds ]
allBoundIds (App e1 e2) = allBoundIds e1 `unionVarSet` allBoundIds e2
allBoundIds (Case scrut _ _ alts) =
    allBoundIds scrut `unionVarSet` unionVarSets
        [ allBoundIds e | (_, _ , e) <- alts ]
allBoundIds (Lam _ e)  = allBoundIds e
allBoundIds (Tick _ e) = allBoundIds e
allBoundIds (Cast e _) = allBoundIds e
allBoundIds _ = emptyVarSet
