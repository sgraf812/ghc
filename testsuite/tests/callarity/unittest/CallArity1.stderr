go2:
    go C(C(A))
    d C(U)
nested_go2:
    go C(C(U))
    go2 C(C(U))
    d C(U)
    n C(U)
d0 (go 2 would be bad):
    go C(U)
    d U
go2 (in case crut):
    go C(C(U))
    d C(U)
go2 (in function call):
    go C(C(U))
    d C(U)
go2 (using surrounding interesting let):
    go C(C(U))
    d C(U)
    n C(U)
go2 (using surrounding boring let):
    go C(C(U))
    d C(U)
    z U
two calls, one from let and from body (d 1 would be bad):
    go C(C(U))
    d U
a thunk in a recursion (d 1 would be bad):
    d U
    n U
two thunks, one called multiple times (both arity 1 would be bad!):
    d U
    n C(U)
two functions, not thunks:
    go C(C(U))
    go2 C(C(U))
a thunk, called multiple times via a forking recursion (d 1 would be bad!):
    go2 C(C(U))
    d U
a function, one called multiple times via a forking recursion:
    go C(C(U))
    go2 C(C(U))
two functions (recursive):
    go C(C(U))
    go2 C(C(U))
mutual recursion (thunks), called mutiple times (both arity 1 would be bad!):
    d U
    n U
mutual recursion (functions), but no thunks:
    go C(C(U))
    go2 C(C(U))
mutual recursion (functions), one boring (d 1 would be bad):
    go C(C(U))
    go2 C(C(U))
    d U
a thunk (non-function-type), called twice, still calls once:
    x U
    d C(U)
a thunk (function type), called multiple times, still calls once:
    d C(U)
    n U
a thunk (non-function-type), in mutual recursion, still calls once (d 0 would be bad):
    go C(C(U))
    x U
    d C(U)
a thunk (non-function-type), in mutual recursion, causes many calls (d 1 would be bad):
    go C(C(U))
    x U
    d U
a thunk (function type), in mutual recursion, still calls once (d 1 would be good):
    go C(U)
    d C(U)
    n U
a thunk (non-function-type) co-calls with the body (d 1 would be bad):
    x U
    d U
